### ISSUE TITLE
Backprop sometimes needs additional memory

### ISSUE THREAD

**Issue by toslunar** [chainer/chainer#5266]
Revisit https://github.com/chainer/chainer/pull/4301#issuecomment-392438444

# Status

I measured how much memory is used in back-propagation.  The peak consumptions of gpu memory (max of `cuda.memory_pool.used_bytes`) are

|model|initialize|forward|backward|fwd/bwd|
|-|-|-|-|-|
|ResNet50|`0127614464`|`3626270720`|`4164088320`|0.8708|
|U-Net|`0081935360`|`0785561088`|`1011115520`|0.7769|

- ResNet50
  - code: `chainer.links.model.vision.resnet.ResNet50Layers`
  - input: `(32, 3, 256, 256), float32`
- U-Net
  - code: `UNET` in https://github.com/pfnet/PaintsChainer/blob/master/cgi-bin/paint_x2_unet/unet.py
  - input: `(16, 4, 128, 128), float32`

# Feature request

Such memory would be reduced by a feature that allows `del loss` before executing `loss.backward()`.  I'd like to implement it.

**Comment by niboshi** [comment]
Could you elaborate what this means?
>`del loss` before executing `loss.backward()`

Is is really related to #4301? #4301 does not reduce maximum memory consumption.

**Comment by toslunar** [comment]
Reopen the issue as a PR is ongoing.
